File: OopsLayout.Windows/UserExceptions.cs
using System.Text.Json;
using OopsLayout.Core;

namespace OopsLayout.Windows;

/// <summary>
/// Persists the user's own keep-words to
/// %AppData%\OopsLayout\exceptions.json and feeds them into the Core keep-lists.
/// Words are routed to the RU or EN list by script. Mirrors UserExceptions.swift.
/// </summary>
internal static class UserExceptions
{
    private sealed class Store
    {
        public List<string> Ru { get; set; } = new();
        public List<string> En { get; set; } = new();
    }

    private static string Dir => Path.Combine(
        Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "OopsLayout");
    private static string FilePath => Path.Combine(Dir, "exceptions.json");

    private static HashSet<string> _ru = new();
    private static HashSet<string> _en = new();

    /// <summary>All user words across both scripts, sorted — for display/editing.</summary>
    public static IEnumerable<string> All => _ru.Concat(_en).OrderBy(w => w, StringComparer.Ordinal);

    public static void Load()
    {
        try
        {
            if (File.Exists(FilePath) &&
                JsonSerializer.Deserialize<Store>(File.ReadAllText(FilePath)) is { } s)
            {
                _ru = new HashSet<string>(s.Ru);
                _en = new HashSet<string>(s.En);
            }
        }
        catch
        {
            // Corrupt/unreadable file — start empty rather than crash.
        }
        Apply();
    }

    /// <summary>
    /// Replace the whole set from a freeform list of words (one per line in the
    /// settings window), routing each by script.
    /// </summary>
    public static void SetWords(IEnumerable<string> words)
    {
        _ru = new HashSet<string>();
        _en = new HashSet<string>();
        foreach (var raw in words)
        {
            var w = raw.Trim().ToLowerInvariant();
            if (w.Length == 0) continue;
            if (w.Any(IsCyrillic)) _ru.Add(w);
            else if (w.Any(char.IsLetter)) _en.Add(w);
        }
        Save();
        Apply();
    }

    private static bool IsCyrillic(char c) => c >= 0x0400 && c <= 0x04FF;

    private static void Apply()
    {
        WordBuffer.UserKeepWordsRu = _ru;
        WordBuffer.UserKeepWordsEn = _en;
    }

    private static void Save()
    {
        try
        {
            Directory.CreateDirectory(Dir);
            var json = JsonSerializer.Serialize(
                new Store { Ru = _ru.ToList(), En = _en.ToList() },
                new JsonSerializerOptions { WriteIndented = true });
            File.WriteAllText(FilePath, json);
        }
        catch
        {
            // Best-effort persistence; never crash the app over it.
        }
    }
}

File: OopsLayout.Mac/Sources/OopsLayout/UserExceptions.swift
import Foundation
import OopsLayoutCore

private extension Character {
    var isCyrillic: Bool {
        unicodeScalars.contains { (0x0400...0x04FF).contains($0.value) }
    }
}

/// Persists the user's own keep-words to
/// ~/Library/Application Support/OopsLayout/exceptions.json and feeds them into
/// the Core keep-lists. Words are routed to the RU or EN list by script.
enum UserExceptions {
    private struct Store: Codable { var ru: [String]; var en: [String] }

    private static var dir: URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("OopsLayout", isDirectory: true)
    }
    private static var fileURL: URL { dir.appendingPathComponent("exceptions.json") }

    private(set) static var ru: Set<String> = []
    private(set) static var en: Set<String> = []

    /// All user words across both scripts, sorted — for display/editing.
    static var all: [String] { ru.union(en).sorted() }

    static func load() {
        if let data = try? Data(contentsOf: fileURL),
           let s = try? JSONDecoder().decode(Store.self, from: data) {
            ru = Set(s.ru)
            en = Set(s.en)
        }
        apply()
    }

    /// Replace the whole set from a freeform list of words (one per line in the
    /// settings window), routing each by script.
    static func setWords(_ words: [String]) {
        ru = []
        en = []
        for raw in words {
            let w = raw.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
            guard !w.isEmpty else { continue }
            if w.contains(where: { $0.isCyrillic }) {
                ru.insert(w)
            } else if w.contains(where: { $0.isLetter }) {
                en.insert(w)
            }
        }
        save()
        apply()
    }

    private static func apply() {
        WordBuffer.userKeepWordsRu = ru
        WordBuffer.userKeepWordsEn = en
    }

    private static func save() {
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        let store = Store(ru: Array(ru), en: Array(en))
        if let data = try? JSONEncoder().encode(store) {
            try? data.write(to: fileURL)
        }
    }
}
